### ISSUE TITLE
Capistrano failed deploy is removing previous releases

### ISSUE THREAD

**Issue by fields** [capistrano/capistrano#1885]
We have a capistrano deployment that is failing from one location because it cannot access the repository on github because of key issues (other users can run the deploy just fine). Whenever this deploy runs, it never gets to the point where it creates a new release directory, but it still sees the release as failed and rolls back the last previous one which was deployed successfully by someone else.

_This should be considered data loss and should never happen._

**Comment by leehambley** [comment]
> This should be considered data loss and should never happen.

Please clarify. If I understood what you're saying we're rolling back to `-2` when we should only be rolling back `-1`?

**Comment by mattbrictson** [comment]
Thanks for opening an issue. Can you help us by giving the version number of Capistrano you are using? If possible, please also post the output of the `cap production doctor` command, with any sensitive information redacted.

**Comment by fields** [comment]
@leehambley the rollback here should just do nothing, because a new release directory was not created as part of this deploy. Instead, it's taking the previous release and rolling _that_ back.

**Comment by fields** [comment]
@mattbrictson This is capistrano 3.7.1. I will get that output for you.

**Comment by mattbrictson** [comment]
To my knowledge Capistrano 3.7 does not automatically rollback on a failed release. Is there a step you didn't mention in your issue description? Are you running `rollback` explicitly? With what arguments? Thanks!

**Comment by fields** [comment]
@mattbrictson Hmmm. I inherited some of this config, and it does have:

```namespace :deploy do
  after :publishing, :restart_unicorn
  after :failed, :rollback
end```

That shouldn't matter though, right? Even in that case, the rollback should still not remove a directory that wasn't created in this deploy.

**Comment by fields** [comment]
There is a lot of sensitive stuff in the variable section for cap doctor. Here is the environment and gems:

```
Environment

    Ruby     ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-linux]
    Rubygems 2.6.11
    Bundler  1.14.6
    Command  /home/maviewer/rails/shared/bundle/ruby/2.3.0/bin/cap production_xxxx doctor

Gems

    capistrano          3.7.1  (update available)
    airbrussh           1.1.2  (update available)
    rake                12.0.0
    sshkit              1.11.5 (update available)
    net-ssh             4.0.1  (update available)
    capistrano-bundler  1.1.4  (update available)
    capistrano-harrow   0.5.3
    capistrano-rails    1.1.6  (update available)
    capistrano-rvm      0.1.2
    capistrano-sidekiq  0.5.4  (update available)
    capistrano3-unicorn 0.2.1
```

If there's something else specific you're looking for, let me know.

**Comment by will-in-wi** [comment]
I'd remove `after :failed, :rollback`. Rollback assumes that the current symlink has been switched to the new release, and (IMO) is usually intended to be run manually.

**Comment by mattbrictson** [comment]
Agreed. `after :failed, :rollback` is a clever hack but it does not work. Capistrano's `rollback` task is [very basic](http://capistranorb.com/documentation/getting-started/rollbacks/) and meant to be run manually as a separate command. It does not know why the deploy failed, which part of the deploy failed, or even which servers failed; it is a very blunt instrument.

In your case, if the git checkout step failed, there is no need to run a rollback at all, because no deploy has actually taken place (where "deploy" is defined as changing the `current` symlink).

There is a case to be made for a rollback feature that is "smarter". But to be honest, I don't think that is something I would put on the roadmap for this project. It is simply too difficult to implement well and without risk of data loss.

Anyway, long story short: I recommend removing the `after` trick. If you want to run `rollback`, do it manually and using command-line options to target the specific host(s) you'd like to affect. My personal approach is to not use `rollback` at all; instead I re-run `deploy` using a known-good commit if a deploy fails.

**Comment by will-in-wi** [comment]
Sounds like this is a configuration issue. Closing.

Feel free to reopen if anyone disagrees.

**Comment by Vanuan** [comment]
As far as I understood, this can be solved using the following:

1. Add task which shows current revision:

```ruby
namespace :deploy do
  desc 'Show deployed release'
  task :show_release do
    on roles(:app) do
        execute "releases=($(ls releases)) ; echo \"${releases[0]}\""
    end
  end
end
```

2. Use a script which rollbacks to the known revision rather than any previous revision:

```bash
CURRENT_RELEASE=`bundle exec cap $end deploy:show_release`
bundle exec cap $env deploy --trace
if [ "$?" -ne 0 ]; then
  set -e
  echo "There was a failure on deploy. Trying to roll back."
  bundle exec cap $env deploy:rollback ROLLBACK_RELEASE=$CURRENT_RELEASE --trace 
  exit 1
fi
```
